[UG-665] Connect deploy node vm as a Jenkins slave

mkam opened the pull request:
This allows commands to be run directly
on the deploy node instead of having
an extra ssh step.

Issue: [UG-665](https://rpc-openstack.atlassian.net/browse/UG-665)

mkam commented:
I didn't remove the ssh to VM code because I thought that should go in a follow up PR in case we still need the functionality or need to do a revert. 

I've had successful builds with master and mitaka, but one thing that isn't working is pulling out the tempest results file because they're in different locations for AIO and multi-node AIO. This doesn't affect whether the build passes or fails from tempest, but we won't have the test results report page. I'll try to update this once https://github.com/rcbops/rpc-gating/pull/295 merges.

hughsaunders commented on rpc_jobs/hw_multi_node_aio.yml:
```diff
@@ -129,8 +130,26 @@
+                include: "inventory/hosts"
+              )
+            }}
+            ssh_slave.connect(2222)
```
If you wrote both nodes to the same inventory, you'd only need to call ssh_slave.connect once? I like that you're connecting both the host and the VM as nodes, provides flexibility.

mkam replied:
The VM creation is still being handled as a step that runs on the host server slave, so when the initial connect happens, the VM doesn't exist.

hughsaunders commented on rpc_jobs/multi_node_aio.yml:
```diff
@@ -186,6 +207,10 @@
             maas.entity_cleanup(instance_name: instance_name)
+            if(deploy_node != null){{
+              instance_name = deploy_node
+              ssh_slave.destroy()
```
Also destroy should take a name param, so we don't have to set a global to different values between two calls of the function.

hughsaunders commented on rpc_jobs/hw_multi_node_aio.yml:
```diff
@@ -139,21 +158,23 @@
               print e
               throw e
           }} finally {{
+              ssh_slave.destroy()
```
its not obvious that these two calls of ssh_slave.destroy() do different things as there are no different parameters. As noted in another comment, I think destroy should be refactored to take a name argument.

mkam commented:
Added the slave_name parameter to the destroy method and went ahead and moved the deploy node connection to its own method.

Run with the latest code: https://rpc.jenkins.cit.rackspace.net/job/mkam-ug-665/30/

mattt416 commented on rpc_jobs/hw_multi_node_aio.yml:
```diff
@@ -139,21 +145,22 @@
+              ssh_slave.destroy(instance_name)
+              maas.entity_cleanup(instance_name: instance_name)
+              if(deploy_node != null){{
+                ssh_slave.destroy(deploy_node)
```
Do we need to check if deploy_node != null?  Would that situation ever exist?

mkam replied:
If the mnaio setup fails, then the deploy node as a slave won't exist, which causes this destroy to fail. It won't change the outcome of the build since it's already failed, but I thought if people saw that failure at a glance, they would think that was the original failure reason.

mattt416 commented on playbooks/setup_jenkins_slave.yml:
```diff
@@ -13,6 +13,7 @@
     # These variables are not booleans so can use the standard default filter
     jenkins_node_labels: "{{ lookup('env', 'JENKINS_NODE_LABELS')|default('single_use_slave', true) }}"
     jenkins_node_executors: "{{ lookup('env', 'JENKINS_NODE_EXECUTORS')|default('2', true) }}"
+    jenkins_port: "{{ ansible_port | default(22, true) }}"
```
The default kind of gives it away, but we could probably indicate w/ the variable name that this is the ssh port we're connecting into.

weezer commented:
LGTM but someone else needs to approve it.
